diff --git a/meson.build b/meson.build
index c044c099..c2ff2d3a 100644
--- a/meson.build
+++ b/meson.build
@@ -917,30 +917,6 @@ test(
         dependencies: test_dep),
     env: test_env,
 )
-if get_option('enable-x11')
-    has_xvfb = find_program('Xvfb', required: false)
-    has_xkbcomp = find_program('xkbcomp', required: false)
-    # We only warn because the build machine may not be the same
-    # as the host/test machine.
-    if not has_xvfb.found()
-        warning('Xvfb program not found, but is required to run X11 tests.')
-    endif
-    if not has_xkbcomp.found()
-        warning('xkbcomp program not found, but is required to run X11 tests.')
-    endif
-    test(
-        'x11',
-        executable('test-x11', 'test/x11.c', dependencies: x11_xvfb_test_dep),
-        env: test_env,
-        is_parallel : false,
-    )
-    test(
-        'x11comp',
-        executable('test-x11comp', 'test/x11comp.c', dependencies: x11_xvfb_test_dep),
-        env: test_env,
-        is_parallel : false,
-    )
-endif
 if get_option('enable-xkbregistry')
     test(
         'registry',
@@ -1077,13 +1053,6 @@ benchmark(
     executable('bench-atom', 'bench/atom.c', dependencies: test_dep),
     env: bench_env,
 )
-if get_option('enable-x11')
-  benchmark(
-      'x11',
-      executable('bench-x11', 'bench/x11.c', dependencies: x11_test_dep),
-      env: bench_env,
-  )
-endif
 
 
 # Documentation.
