diff --git a/src/runtime/Makefile b/src/runtime/Makefile
index 1776c19..6207731 100644
--- a/src/runtime/Makefile
+++ b/src/runtime/Makefile
@@ -1,12 +1,12 @@
 GIT_COMMIT := $(shell cat version)
-CC            = clang
-CFLAGS        = -std=gnu99 -Os -D_FILE_OFFSET_BITS=64 -DGIT_COMMIT=\"$(GIT_COMMIT)\" -T data_sections.ld -ffunction-sections -fdata-sections -Wl,--gc-sections -static -Wall -Werror -static-pie
-LIBS          = -lsquashfuse -lsquashfuse_ll -lzstd -lz -lfuse3
+CC            = gcc
+CFLAGS        = -std=gnu99 -Os -D_FILE_OFFSET_BITS=64 -DGIT_COMMIT=\"$(GIT_COMMIT)\" -T data_sections.ld -ffunction-sections -fdata-sections -Wl,--gc-sections -static -Wall
+LIBS          = -lsquashfuse -lsquashfuse_ll -lzstd -lfuse3 -lpthread
 
 all: runtime
 
 runtime: runtime.c
-	$(CC) -I/usr/local/include/squashfuse -I/usr/include/fuse3 $(CFLAGS) $^ $(LIBS) -o $@
+	$(CC) -L${host_prefix}/lib -I${host_prefix}/include -I${host_prefix}/squashfuse -I${host_prefix}/include/fuse3 $(CFLAGS) $^ $(LIBS) -o $@
 
 clean:
 	rm -f runtime
