diff --git a/src/runtime/Makefile b/src/runtime/Makefile
index d0cbf1f..905b130 100644
--- a/src/runtime/Makefile
+++ b/src/runtime/Makefile
@@ -1,21 +1,21 @@
 CC            = gcc
 CFLAGS        = -std=gnu99 -s -Os -D_FILE_OFFSET_BITS=64 -DGIT_COMMIT=\"${GIT_COMMIT}\" -T data_sections.ld -ffunction-sections -fdata-sections -Wl,--gc-sections -static
-LIBS          = -lsquashfuse -lsquashfuse_ll -lzstd -lz
+LIBS          = -lsquashfuse -lsquashfuse_ll -lzstd -lpthread
 
 all: runtime-fuse3 runtime-fuse3
 
 # Compile runtime
 runtime-fuse3.o: runtime.c
-	$(CC) -I/usr/local/include/squashfuse -I/usr/include/fuse -o runtime-fuse3.o -c $(CFLAGS) $^
+	$(CC) -I${host_prefix}/include -I${host_prefix}/include/squashfuse -I${host_prefix}/include/fuse -o runtime-fuse3.o -c $(CFLAGS) $^
 
 runtime-fuse3: runtime-fuse3.o
 	$(CC) $(CFLAGS) $^ $(LIBS) -lfuse -o runtime-fuse3
 
 runtime-fuse3.o: runtime.c
-	$(CC) -I/usr/local/include/squashfuse -I/usr/include/fuse3 -o runtime-fuse3.o -c $(CFLAGS) $^
+	$(CC) -I${host_prefix}/include -I${host_prefix}/include/squashfuse -I${host_prefix}/include/fuse3 -o runtime-fuse3.o -c $(CFLAGS) $^
 
 runtime-fuse3: runtime-fuse3.o
-	$(CC) $(CFLAGS) $^ $(LIBS) -lfuse3 -o runtime-fuse3
+	$(CC) $(CFLAGS) $^ -L${host_prefix}/lib $(LIBS) -lfuse3 -lpthread -o runtime-fuse3
 
 clean:
 	rm -f *.o runtime-fuse3 runtime-fuse3
