diff --git a/src/plugins/platforms/cocoa/qcocoaintegration.h b/src/plugins/platforms/cocoa/qcocoaintegration.h
index 256b7b36ad..c6d73e3c06 100644
--- a/src/plugins/platforms/cocoa/qcocoaintegration.h
+++ b/src/plugins/platforms/cocoa/qcocoaintegration.h
@@ -90,7 +90,6 @@ public:
     void clearToolbars();
 
     void setApplicationIcon(const QIcon &icon) const override;
-    void setApplicationBadge(qint64 number) override;
 
     void beep() const override;
     void quit() const override;
diff --git a/src/plugins/platforms/cocoa/qcocoaintegration.mm b/src/plugins/platforms/cocoa/qcocoaintegration.mm
index 2ec225cbea..b17ab38b10 100644
--- a/src/plugins/platforms/cocoa/qcocoaintegration.mm
+++ b/src/plugins/platforms/cocoa/qcocoaintegration.mm
@@ -438,11 +438,6 @@ void QCocoaIntegration::setApplicationIcon(const QIcon &icon) const
     NSApp.applicationIconImage = [NSImage imageFromQIcon:icon withSize:fallbackSize];
 }
 
-void QCocoaIntegration::setApplicationBadge(qint64 number)
-{
-    NSApp.dockTile.badgeLabel = number ? [NSString stringWithFormat:@"%" PRId64, number] : nil;
-}
-
 void QCocoaIntegration::beep() const
 {
     NSBeep();
