diff --git a/cmake/QtBaseHelpers.cmake b/cmake/QtBaseHelpers.cmake
index 4b48f8f176..63e445e346 100644
--- a/cmake/QtBaseHelpers.cmake
+++ b/cmake/QtBaseHelpers.cmake
@@ -212,7 +212,7 @@ macro(qt_internal_qtbase_build_repo)
     endif()
 
     if(NOT QT_INTERNAL_BUILD_STANDALONE_PARTS)
-        if(QT_FEATURE_settings)
+        if(QT_WILL_BUILD_TOOLS AND QT_FEATURE_settings)
             add_subdirectory(qmake)
         endif()
 
diff --git a/qmake/CMakeLists.txt b/qmake/CMakeLists.txt
index 9334ca72ab..801b4f7647 100644
--- a/qmake/CMakeLists.txt
+++ b/qmake/CMakeLists.txt
@@ -112,19 +112,6 @@ qt_internal_add_tool(${target_name}
         generators/win32/msvc_objectmodel.cpp # xml symbols clash with generators/win32/msbuild_objectmodel.cpp
         library/qmakeparser.cpp               # struct statics clashes with qmakeevaluator_p.h
 )
-
-# When we don't build tools, we still want to be able to build the documentation.
-# Add a fake interface library to make sure the documentation target is created.
-# We use the direct name rather than ${target_name} to ensure the doc target keeps
-# the original name even when cross-building.
-if(NOT TARGET qmake)
-    add_library(qmake INTERFACE)
-endif()
-
-qt_internal_add_docs(qmake
-    doc/qmake.qdocconf
-)
-
 qt_internal_return_unless_building_tools()
 
 #### Keys ignored in scope 1:.:.:qmake.pro:<TRUE>:
@@ -154,3 +141,6 @@ set_target_properties(${target_name} PROPERTIES
 qt_internal_apply_gc_binaries(${target_name} PRIVATE)
 qt_skip_warnings_are_errors(${target_name})
 
+qt_internal_add_docs(${target_name}
+    doc/qmake.qdocconf
+)
-- 
2.46.2

