diff --git a/src/gui/platform/darwin/qappleiconengine.mm b/src/gui/platform/darwin/qappleiconengine.mm
index a346c4d1f1..a1d02a97cb 100644
--- a/src/gui/platform/darwin/qappleiconengine.mm
+++ b/src/gui/platform/darwin/qappleiconengine.mm
@@ -362,15 +362,6 @@ auto *configuredImage(const NSImage *image, const QColor &color)
     auto *config = [NSImageSymbolConfiguration configurationWithPointSize:48
                                                weight:NSFontWeightRegular
                                                scale:NSImageSymbolScaleLarge];
-    if (@available(macOS 12, *)) {
-        auto *primaryColor = [NSColor colorWithSRGBRed:color.redF()
-                                                 green:color.greenF()
-                                                  blue:color.blueF()
-                                                 alpha:color.alphaF()];
-
-        auto *colorConfig = [NSImageSymbolConfiguration configurationWithHierarchicalColor:primaryColor];
-        config = [config configurationByApplyingConfiguration:colorConfig];
-    }
 
     return [image imageWithSymbolConfiguration:config];
 }
diff --git a/src/gui/rhi/qrhimetal.mm b/src/gui/rhi/qrhimetal.mm
index 232b2374b3..d28d966ad1 100644
--- a/src/gui/rhi/qrhimetal.mm
+++ b/src/gui/rhi/qrhimetal.mm
@@ -4685,10 +4685,6 @@ id<MTLLibrary> QRhiMetalData::createMetalLib(const QShader &shader, QShader::Var
                                              QString *error, QByteArray *entryPoint, QShaderKey *activeKey)
 {
     QVarLengthArray<int, 8> versions;
-    if (@available(macOS 13, iOS 16, *))
-        versions << 30;
-    if (@available(macOS 12, iOS 15, *))
-        versions << 24;
     if (@available(macOS 11, iOS 14, *))
         versions << 23;
     if (@available(macOS 10.15, iOS 13, *))
