diff --git a/src/plugins/multimedia/darwin/camera/qavfcamerabase.mm b/src/plugins/multimedia/darwin/camera/qavfcamerabase.mm
index 940a964b7..bab2cebe0 100644
--- a/src/plugins/multimedia/darwin/camera/qavfcamerabase.mm
+++ b/src/plugins/multimedia/darwin/camera/qavfcamerabase.mm
@@ -176,14 +176,6 @@ void QAVFVideoDevices::updateCameraDevices()
 #endif
     ];
 
-#if QT_DARWIN_PLATFORM_SDK_EQUAL_OR_ABOVE(__MAC_14_0, __IPHONE_17_0, __TVOS_NA, __WATCHOS_NA)
-    if (@available(macOS 14, iOS 17, *)) {
-      discoveryDevices = [discoveryDevices arrayByAddingObjectsFromArray: @[
-        AVCaptureDeviceTypeExternal,
-        AVCaptureDeviceTypeContinuityCamera
-      ]];
-    } else
-#endif
     {
 #ifdef Q_OS_MACOS
     QT_WARNING_PUSH
