diff --git a/src/gui/platform/darwin/qappleiconengine.mm b/src/gui/platform/darwin/qappleiconengine.mm
index d99da6da84..c6cd8ffe87 100644
--- a/src/gui/platform/darwin/qappleiconengine.mm
+++ b/src/gui/platform/darwin/qappleiconengine.mm
@@ -362,15 +362,6 @@ auto *configuredImage(const NSImage *image, const QColor &color)
     auto *config = [NSImageSymbolConfiguration configurationWithPointSize:48
                                                weight:NSFontWeightRegular
                                                scale:NSImageSymbolScaleLarge];
-    if (@available(macOS 12, *)) {
-        auto *primaryColor = [NSColor colorWithSRGBRed:color.redF()
-                                                 green:color.greenF()
-                                                  blue:color.blueF()
-                                                 alpha:color.alphaF()];
-
-        auto *colorConfig = [NSImageSymbolConfiguration configurationWithHierarchicalColor:primaryColor];
-        config = [config configurationByApplyingConfiguration:colorConfig];
-    }
 
     return [image imageWithSymbolConfiguration:config];
 }
diff --git a/src/gui/rhi/qrhimetal.mm b/src/gui/rhi/qrhimetal.mm
index 3046b76427..8c1e7fd13f 100644
--- a/src/gui/rhi/qrhimetal.mm
+++ b/src/gui/rhi/qrhimetal.mm
@@ -4710,10 +4710,6 @@ id<MTLLibrary> QRhiMetalData::createMetalLib(const QShader &shader, QShader::Var
                                              QString *error, QByteArray *entryPoint, QShaderKey *activeKey)
 {
     QVarLengthArray<int, 8> versions;
-    if (@available(macOS 13, iOS 16, *))
-        versions << 30;
-    if (@available(macOS 12, iOS 15, *))
-        versions << 24;
     versions << 23 << 22 << 21 << 20 << 12;
 
     const QList<QShaderKey> shaders = shader.availableShaders();
