diff --git a/src/gui/rhi/qshader_p.h b/src/gui/rhi/qshader_p.h
index 690a7f44cd..8e5894c97f 100644
--- a/src/gui/rhi/qshader_p.h
+++ b/src/gui/rhi/qshader_p.h
@@ -17,6 +17,7 @@
 
 #include <QtGui/qtguiglobal.h>
 #include <QtCore/qhash.h>
+#include <QtCore/qmap.h>
 #include <private/qshaderdescription_p.h>
 
 QT_BEGIN_NAMESPACE
@@ -127,7 +128,7 @@ public:
     QByteArray serialized() const;
     static QShader fromSerialized(const QByteArray &data);
 
-    using NativeResourceBindingMap = QHash<int, QPair<int, int> >; // binding -> native_binding[, native_binding]
+    using NativeResourceBindingMap = QMap<int, QPair<int, int> >; // binding -> native_binding[, native_binding]
     NativeResourceBindingMap nativeResourceBindingMap(const QShaderKey &key) const;
     void setResourceBindingMap(const QShaderKey &key, const NativeResourceBindingMap &map);
     void removeResourceBindingMap(const QShaderKey &key);
