diff --git a/src/plugins/multimedia/windows/player/mfplayersession.cpp b/src/plugins/multimedia/windows/player/mfplayersession.cpp
index ccdd5b6c5..94218a5b9 100644
--- a/src/plugins/multimedia/windows/player/mfplayersession.cpp
+++ b/src/plugins/multimedia/windows/player/mfplayersession.cpp
@@ -30,7 +30,7 @@
 #include <mfidl.h>
 #include <mmdeviceapi.h>
 #include <propvarutil.h>
-#include <Wininet.h>
+#include <wininet.h>
 #include <functiondiscoverykeys_devpkey.h>
 
 //#define DEBUG_MEDIAFOUNDATION
