diff --git a/src/widgets/CMakeLists.txt b/src/widgets/CMakeLists.txt
index aaa81286ab..39c197eeba 100644
--- a/src/widgets/CMakeLists.txt
+++ b/src/widgets/CMakeLists.txt
@@ -693,7 +693,7 @@ if(QT_FEATURE_wizard)
     if(APPLE)
         set_source_files_properties(dialogs/qwizard.cpp
             PROPERTIES
-                LANGUAGE OBJCXX
+                COMPILE_FLAGS "-x objective-c++"
                 SKIP_PRECOMPILE_HEADERS ON
         )
     endif()
diff --git a/tests/auto/corelib/io/qdebug/CMakeLists.txt b/tests/auto/corelib/io/qdebug/CMakeLists.txt
index c9fd9c6df8..cef57a2186 100644
--- a/tests/auto/corelib/io/qdebug/CMakeLists.txt
+++ b/tests/auto/corelib/io/qdebug/CMakeLists.txt
@@ -13,5 +13,5 @@ qt_internal_add_test(tst_qdebug
 )

 if (APPLE)
-    set_source_files_properties(tst_qdebug.cpp PROPERTIES LANGUAGE OBJCXX)
+    target_compile_options(tst_qdebug PRIVATE -x objective-c++)
 endif()
diff --git a/tests/auto/corelib/io/qfileinfo/CMakeLists.txt b/tests/auto/corelib/io/qfileinfo/CMakeLists.txt
index 4da51ce219..5fdc764e8a 100644
--- a/tests/auto/corelib/io/qfileinfo/CMakeLists.txt
+++ b/tests/auto/corelib/io/qfileinfo/CMakeLists.txt
@@ -48,5 +48,5 @@ qt_internal_extend_target(tst_qfileinfo CONDITION WIN32
 )

 if (APPLE)
-    set_source_files_properties(tst_qfileinfo.cpp PROPERTIES LANGUAGE OBJCXX)
+    target_compile_options(tst_qfileinfo PRIVATE -x objective-c++)
 endif()
diff --git a/tests/auto/gui/kernel/qwindow/CMakeLists.txt b/tests/auto/gui/kernel/qwindow/CMakeLists.txt
index 5ddda80c2a..605761622f 100644
--- a/tests/auto/gui/kernel/qwindow/CMakeLists.txt
+++ b/tests/auto/gui/kernel/qwindow/CMakeLists.txt
@@ -25,7 +25,7 @@ if(APPLE OR WIN32)
     )

     if(APPLE)
-        set_source_files_properties(tst_foreignwindow.cpp PROPERTIES LANGUAGE OBJCXX)
+        target_compile_options(tst_foreignwindow PRIVATE -x objective-c++)
         set_property(TARGET tst_foreignwindow PROPERTY PROPERTY MACOSX_BUNDLE TRUE)
     endif()
 endif()
--
2.40.1

