From 13214d18c6ea1671dcff3dd62ccf88007b41f112 Mon Sep 17 00:00:00 2001
From: Frank Denis <github@pureftpd.org>
Date: Wed, 1 Jul 2020 15:33:52 +0200
Subject: [PATCH] Recognize the arm + apple combo as aarch64

---
 configure.ac | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/configure.ac b/configure.ac
index 05a51fa5..3993dbc1 100644
--- a/configure.ac
+++ b/configure.ac
@@ -214,6 +214,8 @@ AX_VALGRIND_CHECK
 
 dnl Checks
 
+AS_IF([test "$host_cpu" = "arm" && test "$host_vendor" = "apple"], [host_cpu=aarch64])
+
 AC_C_VARARRAYS
 
 AC_CHECK_DEFINE([__wasi__], [WASI="yes"], [])
-- 
2.38.1

