diff --git a/qtbase/configure b/qtbase/configure
index e57707dc..d8f6c934 100755
--- a/qtbase/configure
+++ b/qtbase/configure
@@ -9,9 +9,9 @@
 
 # the directory of this script is the "source tree"
 relpath=`dirname "$0"`
-relpath=`(cd "$relpath"; /bin/pwd)`
+relpath=`(cd "$relpath"; pwd)`
 # the current directory is the "build tree" or "object tree"
-outpath=`/bin/pwd`
+outpath=`pwd`
 outpathPrefix=$outpath
 
 # do this early so we don't store it in config.status
