--- a/lib/fuse.c
+++ b/lib/fuse.c
@@ -227,32 +227,6 @@ static struct fuse_module *fuse_modules;
 
 static int fuse_load_so_name(const char *soname)
 {
-	struct fusemod_so *so;
-
-	so = calloc(1, sizeof(struct fusemod_so));
-	if (!so) {
-		fprintf(stderr, "fuse: memory allocation failed\n");
-		return -1;
-	}
-
-	fuse_current_so = so;
-	so->handle = dlopen(soname, RTLD_NOW);
-	fuse_current_so = NULL;
-	if (!so->handle) {
-		fprintf(stderr, "fuse: %s\n", dlerror());
-		goto err;
-	}
-	if (!so->ctr) {
-		fprintf(stderr, "fuse: %s did not register any modules\n",
-			soname);
-		goto err;
-	}
-	return 0;
-
-err:
-	if (so->handle)
-		dlclose(so->handle);
-	free(so);
 	return -1;
 }
 
@@ -314,7 +288,6 @@ static void fuse_put_module(struct fuse_module *m)
 				else
 					mp = &(*mp)->next;
 			}
-			dlclose(so->handle);
 			free(so);
 		}
 	}
-- 
2.40.0

