diff --git a/lib/meson.build b/lib/meson.build
index 9044630..ea09d45 100644
--- a/lib/meson.build
+++ b/lib/meson.build
@@ -32,7 +32,7 @@ else
    deps += cc.find_library('rt')
 endif
 
-fusermount_path = join_paths(get_option('prefix'), get_option('bindir'))
+fusermount_path = '/usr/bin'
 libfuse = library('fuse3', libfuse_sources, version: meson.project_version(),
                   soversion: '3', include_directories: include_dirs,
                   dependencies: deps, install: true,
